py/server/deephaven/pandas.py

```diff
@@ -173,20 +202,27 @@ def to_pandas(table: Table, cols: List[str] = None, dtype_backend: str = None) -
 }
 
 
-def _map_na(np_array: np.ndarray):
+def _map_na(array: [np.ndarray, pd.api.extensions.ExtensionArray]):
     """Replaces the pd.NA values in the array if it is of pandas ExtensionDtype(nullable)."""
-    pd_dtype = np_array.dtype
+    pd_dtype = array.dtype
     if not isinstance(pd_dtype, pd.api.extensions.ExtensionDtype):
-        return np_array
+        return array
 
     dh_null = _EX_DTYPE_NULL_MAP.get(type(pd_dtype)) or _EX_DTYPE_NULL_MAP.get(pd_dtype)
-    if isinstance(pd_dtype, pd.StringDtype) or isinstance(pd_dtype, pd.BooleanDtype) or pd_dtype == pd.ArrowDtype(
-            pa.bool_()):
-        np_array = np.array(list(map(lambda v: dh_null if v is pd.NA else v, np_array)))
+    # To preserve NaNs in floating point arrays, Pandas doesn't distinguish NaN/Null as far as NA testing is
+    # concerted, thus its fillna() method will replace both NaN/Null in the data.
+    if isinstance(pd_dtype, (pd.Float32Dtype, pd.Float64Dtype)) and isinstance(getattr(array, "_data"), np.ndarray):
+        np_array = array._data
+        null_mask = np.logical_and(array._mask, np.logical_not(np.isnan(np_array)))
+        np_array[null_mask] = dh_null
+        return np_array
```

Comment:
Should there be a function argument indicating if `NaN` gets converted as `NaN` or `NULL`?

Reply:
See the comment above about supporting NaN conversion.

Reply:
It looks like this converts arrays with `NaN` and `NULL` values to arrays with `dh_null` values in their place. Is this not correct? If so, why worry about converting things to `NaN`?

Reply:
The behavior that matters in this case is what happens when a user would covert a table to a dataframe and back. If the table has nulls and NaNs, I would expect the table before and after the conversion to be identical unless that's just not possible.